src/legacy/ui/public/agg_types/controls/sub_metric.tsx

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React, { useEffect, useState } from 'react';
+import { EuiFormLabel } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { AggParamEditorProps, DefaultEditorAggParams } from '../../vis/editors/default';
+import { AggConfig } from '../../vis';
+
+function SubMetricParamEditor({
+  agg,
+  aggParam,
+  value,
+  responseValueAggs,
+  state,
+  setValue,
+  setValidity,
+  setTouched,
+  subAggParams,
+}: AggParamEditorProps<AggConfig>) {
+  const metricTitle = i18n.translate('common.ui.aggTypes.metrics.metricTitle', {
+    defaultMessage: 'Metric',
+  });
+  const bucketTitle = i18n.translate('common.ui.aggTypes.metrics.bucketTitle', {
+    defaultMessage: 'Bucket',
+  });
+  const type = aggParam.name;
+
+  const aggTitle = type === 'customMetric' ? metricTitle : bucketTitle;
+  const aggGroup = type === 'customMetric' ? 'metrics' : 'buckets';
+
+  // reset validity before component destroyed
+  useEffect(
```

Comment:
That is actually a rather useful  pattern. Let's make this a custom hook `useUnmount` where we can pass in that function directly:

```ts
export function useUnmount(fn: () => void) {
  useEffect(() => fn, []);
}
```

We already have a plugin for react utilities, where this could perfectly go: `src/plugins/kibana_react/public/util/use_unmount.ts`

Reply:
Thanks. I added a custom hook.